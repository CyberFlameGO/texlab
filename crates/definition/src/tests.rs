use crate::{DefinitionParams, DefinitionResult};

fn check(input: &str) {
    let fixture = test_utils::fixture::Fixture::parse(input);
    let workspace = &fixture.workspace;

    let mut origin_selection_range = None;
    let mut origin_document = None;
    let mut origin_cursor = None;
    for document in &fixture.documents {
        if let Some(cursor) = document.cursor {
            origin_document = Some(document);
            origin_cursor = Some(cursor);
            origin_selection_range = document
                .ranges
                .iter()
                .find(|range| range.contains_inclusive(cursor))
                .copied();

            break;
        }
    }

    let origin_document = origin_document.unwrap();

    let mut expected = Vec::new();
    for document in &fixture.documents {
        let mut ranges = document.ranges.iter();
        while let Some(target_selection_range) = ranges.next().copied() {
            let origin_selection_range = origin_selection_range.unwrap();
            if (&origin_document.uri, origin_selection_range)
                != (&document.uri, target_selection_range)
            {
                expected.push(DefinitionResult {
                    origin_selection_range,
                    target: fixture.workspace.lookup(&document.uri).unwrap(),
                    target_range: *ranges.next().unwrap(),
                    target_selection_range,
                });
            }
        }
    }

    let mut actual = crate::goto_definition(DefinitionParams {
        workspace,
        document: workspace.lookup(&origin_document.uri).unwrap(),
        offset: origin_cursor.unwrap(),
    });

    sort_results(&mut expected);
    sort_results(&mut actual);

    assert_eq!(actual, expected);
}

fn sort_results(items: &mut Vec<DefinitionResult>) {
    items.sort_by(|a, b| {
        let a = (&a.target.uri, a.target_range.start());
        let b = (&b.target.uri, b.target_range.start());
        a.cmp(&b)
    });
}

#[test]
fn test_command_definition() {
    check(
        r#"
%! main.tex
\DeclareMathOperator{\foo}{foo}
                     ^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
\foo
  |
^^^^"#,
    )
}

#[test]
fn test_document() {
    check(
        r#"
%! foo.tex
\addbibresource{baz.bib}
                  |
                ^^^^^^^

%! bar.bib
@article{foo, bar = {baz}}

%! baz.bib
@article{foo, bar = {baz}}
!
!"#,
    )
}

#[test]
fn test_entry() {
    check(
        r#"
%! foo.tex
\addbibresource{baz.bib}
\cite{foo}
      |
      ^^^

%! bar.bib
@article{foo, bar = {baz}}

%! baz.bib
@article{foo, bar = {baz}}
         ^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^"#,
    )
}

#[test]
fn test_string_simple() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
        ^^^
^^^^^^^^^^^^^^^^^^^^
@article{bar, author = foo}
                        |
                       ^^^"#,
    )
}

#[test]
fn test_string_join() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
        ^^^
^^^^^^^^^^^^^^^^^^^^
@article{bar, author = foo # "bar"}
                        |
                       ^^^"#,
    )
}

#[test]
fn test_string_field() {
    check(
        r#"
%! main.bib
@string{foo = {bar}}
@article{bar, author = foo # "bar"}
                |"#,
    )
}
