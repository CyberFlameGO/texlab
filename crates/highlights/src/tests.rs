use crate::{HighlightKind, HighlightParams};

fn check(input: &str, expected_kinds: &[HighlightKind]) {
    let fixture = test_utils::fixture::Fixture::parse(input);

    let (feature, offset) = fixture.make_params().unwrap();

    let expected_ranges = fixture
        .locations()
        .map(|location| location.range)
        .collect::<Vec<_>>();

    let results = crate::find_all(HighlightParams { feature, offset });

    let actual_ranges = results
        .iter()
        .map(|result| result.range)
        .collect::<Vec<_>>();
    assert_eq!(actual_ranges, expected_ranges);

    assert_eq!(
        results.iter().map(|result| result.kind).collect::<Vec<_>>(),
        expected_kinds
    );
}

#[test]
fn test_label() {
    check(
        r#"
%! main.tex
\label{foo}
        |
       ^^^
\ref{foo}
     ^^^
\label{bar}
"#,
        &[HighlightKind::Write, HighlightKind::Read],
    )
}
