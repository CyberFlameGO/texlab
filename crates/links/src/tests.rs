use expect_test::{expect, Expect};

fn check(input: &str, expect: Expect) {
    let fixture = test_utils::fixture::Fixture::parse(input);
    let (params, _) = fixture.make_params().unwrap();
    let links = crate::find_links(&params);

    let actual_ranges = links.iter().map(|link| link.range).collect::<Vec<_>>();

    let expected_ranges = fixture
        .locations()
        .map(|location| location.range)
        .collect::<Vec<_>>();

    assert_eq!(actual_ranges, expected_ranges);

    let actual_targets = links
        .iter()
        .map(|link| link.document.uri.as_str())
        .collect::<Vec<_>>();

    expect.assert_debug_eq(&actual_targets);
}

#[test]
fn test_document_include() {
    check(
        r#"
%! foo.tex
\input{bar.tex}
       ^^^^^^^

%! bar.tex"#,
        expect![[r#"
            [
                "file:///texlab/bar.tex",
            ]
        "#]],
    );
}

#[test]
fn test_document_import() {
    check(
        r#"
%! foo.tex
\import{.}{bar/baz}
           ^^^^^^^

%! bar/baz.tex"#,
        expect![[r#"
            [
                "file:///texlab/bar/baz.tex",
            ]
        "#]],
    );
}
