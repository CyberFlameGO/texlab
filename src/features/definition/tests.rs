use insta::{
    assert_json_snapshot,
    internals::{Content, ContentPath},
};
use lsp_types::Url;

use crate::features::{goto_definition, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_empty_bibtex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_command_definition() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.tex",
                r#"
\DeclareMathOperator{\foo}{foo}
\foo"#,
            )])
            .main("main.tex")
            .line(1)
            .character(2)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_document_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\addbibresource{baz.bib}"#),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(18)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_entry_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                (
                    "foo.tex",
                    r#"
\addbibresource{baz.bib}
\cite{foo}"#,
                ),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(1)
            .character(6)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_concat() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo # "bar"}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_field() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(18)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

fn redact_uri(value: Content, _path: ContentPath) -> String {
    value.as_str().unwrap().replace(
        Url::from_directory_path(std::env::temp_dir())
            .unwrap()
            .as_str(),
        "[tmp]/",
    )
}
