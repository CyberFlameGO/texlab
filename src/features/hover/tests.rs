use insta::assert_json_snapshot;

use crate::features::{find_hover, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_empty_bibtex_document() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_citation_inside_cite() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![
                (
                    "main.bib",
                    "@article{foo, author = {Foo Bar}, title = {Baz Qux}, year = 1337}",
                ),
                (
                    "main.tex",
                    r#"
\addbibresource{main.bib}
\cite{foo}"#,
                ),
            ])
            .main("main.tex")
            .line(1)
            .character(7)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_citation_inside_entry() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![
                (
                    "main.bib",
                    "@article{foo, author = {Foo Bar}, title = {Baz Qux}, year = 1337}",
                ),
                (
                    "main.tex",
                    r#"
\addbibresource{main.bib}
\cite{foo}"#,
                ),
            ])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_component_known_package() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\usepackage{amsmath}"#)])
            .main("main.tex")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_component_unknown_class() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\documentclass{abcdefghijklmnop}"#)])
            .main("main.tex")
            .line(0)
            .character(20)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_entry_type_known_type() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo,}"#)])
            .main("main.bib")
            .line(0)
            .character(3)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_entry_type_unknown_field() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@foo{bar,}"#)])
            .main("main.bib")
            .line(0)
            .character(3)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_entry_type_key() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo,}"#)])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_field_known() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, author = bar}"#)])
            .main("main.bib")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_field_unknown() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, bar = baz}"#)])
            .main("main.bib")
            .line(0)
            .character(15)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_field_entry_key() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.bib", r#"@article{foo, author = bar}"#)])
            .main("main.bib")
            .line(0)
            .character(11)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}

#[test]
fn test_section() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![("main.tex", r#"\section{Foo}\label{sec:foo}"#)])
            .main("main.tex")
            .line(0)
            .character(23)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_string_inside_reference() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = "Foo"}
@string{bar = "Bar"}
@article{baz, author = bar}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(24)
            .build()
            .hover(),
    );

    assert_json_snapshot!(hover);
}

#[test]
fn test_string_inside_field() {
    let hover = find_hover(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = "Foo"}
@string{bar = "Bar"}
@article{baz, author = bar}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(21)
            .build()
            .hover(),
    );

    assert_eq!(hover, None);
}
