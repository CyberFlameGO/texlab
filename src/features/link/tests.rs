use insta::assert_json_snapshot;

use crate::features::{find_document_links, redactions::redact_uri, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .build()
            .link(),
    );

    assert_eq!(links, Vec::new());
}

#[test]
fn test_empty_bibtex_document() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .build()
            .link(),
    );

    assert_eq!(links, Vec::new());
}

#[test]
fn test_includes() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![("foo.tex", r#"\input{bar.tex}"#), ("bar.tex", r#""#)])
            .main("foo.tex")
            .build()
            .link(),
    );

    assert_json_snapshot!(
        links,
        {
            "[].target" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_imports() {
    let links = find_document_links(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\import{bar/}{baz}"#),
                ("bar/baz.tex", r#""#),
            ])
            .main("foo.tex")
            .build()
            .link(),
    );

    assert_json_snapshot!(
        links,
        {
            "[].target" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}
