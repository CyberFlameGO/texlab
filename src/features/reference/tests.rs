use insta::assert_json_snapshot;

use crate::features::{find_all_references, redactions::redact_uri, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .reference(),
    );

    assert_eq!(references, Vec::new());
}

#[test]
fn test_empty_bibtex_document() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .reference(),
    );

    assert_eq!(references, Vec::new());
}

#[test]
fn test_entry_definition() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.bib", r#"@article{foo,}"#),
                ("bar.tex", r#"\cite{foo}\addbibresource{foo.bib}"#),
            ])
            .main("foo.bib")
            .line(0)
            .character(11)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_entry_definition_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.bib", r#"@article{foo,}"#),
                ("bar.tex", r#"\cite{foo}\addbibresource{foo.bib}"#),
            ])
            .main("foo.bib")
            .line(0)
            .character(11)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_entry_reference() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.bib", r#"@article{foo,}"#),
                ("bar.tex", r#"\cite{foo}\addbibresource{foo.bib}"#),
            ])
            .main("bar.tex")
            .line(0)
            .character(8)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_entry_reference_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.bib", r#"@article{foo,}"#),
                ("bar.tex", r#"\cite{foo}\addbibresource{foo.bib}"#),
            ])
            .main("bar.tex")
            .line(0)
            .character(6)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label_definition() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}"#),
                ("bar.tex", r#"\ref{foo}\input{foo.tex}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(8)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label_definition_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}\input{bar.tex}"#),
                ("bar.tex", r#"\ref{foo}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(9)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label_reference() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}\input{bar.tex}"#),
                ("bar.tex", r#"\ref{foo}"#),
                ("baz.tex", r#"\ref{foo}\input{bar.tex}"#),
            ])
            .main("bar.tex")
            .line(0)
            .character(7)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label_reference_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}"#),
                ("bar.tex", r#"\ref{foo}\input{foo.tex}"#),
            ])
            .main("bar.tex")
            .line(0)
            .character(7)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_string_definition() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {Foo}}
@string{bar = {Bar}}
@article{baz, author = foo}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(24)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_string_definition_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {Foo}}
@string{bar = {Bar}}
@article{baz, author = foo}"#,
            )])
            .main("main.bib")
            .line(2)
            .character(24)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_string_reference() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {Foo}}
@string{bar = {Bar}}
@article{baz, author = foo}"#,
            )])
            .main("main.bib")
            .line(0)
            .character(10)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_string_reference_include_declaration() {
    let references = find_all_references(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {Foo}}
@string{bar = {Bar}}
@article{baz, author = foo}"#,
            )])
            .main("main.bib")
            .line(0)
            .character(10)
            .include_declaration(true)
            .build()
            .reference(),
    );

    assert_json_snapshot!(
        references,
        {
            "[].targetUri" => insta::dynamic_redaction(redact_uri),
            "." => insta::sorted_redaction(),
        }
    );
}
