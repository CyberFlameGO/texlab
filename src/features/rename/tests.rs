use insta::assert_json_snapshot;

use crate::features::{redactions::redact_uri, testing::FeatureTester};

use super::rename_all;

#[test]
fn test_command() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\baz\include{bar.tex}"#),
                ("bar.tex", r#"\baz"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(2)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_entry() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("main.bib", r#"@article{foo, bar = baz}"#),
                ("main.tex", "\\addbibresource{main.bib}\n\\cite{foo}"),
            ])
            .main("main.bib")
            .line(0)
            .character(9)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_citation() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("main.bib", r#"@article{foo, bar = baz}"#),
                ("main.tex", "\\addbibresource{main.bib}\n\\cite{foo}"),
            ])
            .main("main.tex")
            .line(1)
            .character(6)
            .new_name("qux")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}

#[test]
fn test_label() {
    let edit = rename_all(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\label{foo}\include{bar}"#),
                ("bar.tex", r#"\ref{foo}"#),
                ("baz.tex", r#"\ref{foo}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(7)
            .new_name("bar")
            .build()
            .rename(),
    );

    assert_json_snapshot!(
        edit,
        {
            ".changes.$key" => insta::dynamic_redaction(redact_uri),
            ".changes" => insta::sorted_redaction(),
        }
    );
}
