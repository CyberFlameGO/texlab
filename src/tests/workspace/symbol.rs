use anyhow::Result;
use insta::{assert_json_snapshot, internals::Redaction};
use lsp_types::{
    request::WorkspaceSymbol, ClientCapabilities, SymbolInformation, Url, WorkspaceSymbolParams,
};

use crate::tests::{client::Client, fixture};

struct SymbolResult {
    actual_symbols: Vec<SymbolInformation>,
    uri_redaction: Redaction,
}

fn find_symbols(fixture: &str, query: &str) -> Result<SymbolResult> {
    let mut client = Client::spawn()?;
    client.initialize(ClientCapabilities::default(), None)?;

    let fixture = fixture::parse(fixture);
    for file in fixture.files {
        client.open(file.name, file.lang, file.text)?;
    }

    let actual_symbols = client
        .request::<WorkspaceSymbol>(WorkspaceSymbolParams {
            query: query.to_string(),
            work_done_progress_params: Default::default(),
            partial_result_params: Default::default(),
        })?
        .unwrap_or_default();

    let result = client.shutdown()?;

    let uri = Url::from_directory_path(result.directory.path()).unwrap();
    let uri_redaction = insta::dynamic_redaction(move |content, _path| {
        content.as_str().unwrap().replace(uri.as_str(), "[tmp]/")
    });

    Ok(SymbolResult {
        actual_symbols,
        uri_redaction,
    })
}

macro_rules! assert_symbols {
    ($result:expr) => {
        let result = $result;
        assert_json_snapshot!(result.actual_symbols, {
            "[].location.uri" => result.uri_redaction,
            "[]" => insta::sorted_redaction()
        });
    };
}

const FIXTURE: &str = r#"
%TEX main.tex
%SRC \documentclass{article}
%SRC \usepackage{caption}
%SRC \usepackage{amsmath}
%SRC \usepackage{amsthm}
%SRC 
%SRC \begin{document}
%SRC 
%SRC \section{Foo}\label{sec:foo}
%SRC 
%SRC \begin{equation}\label{eq:foo}
%SRC     Foo
%SRC \end{equation}
%SRC 
%SRC \section{Bar}\label{sec:bar}
%SRC 
%SRC \begin{figure}
%SRC     Bar
%SRC     \caption{Bar}
%SRC     \label{fig:bar}
%SRC \end{figure}
%SRC 
%SRC \section{Baz}\label{sec:baz}
%SRC 
%SRC \begin{enumerate}
%SRC     \item\label{itm:foo} Foo
%SRC     \item\label{itm:bar} Bar
%SRC     \item\label{itm:baz} Baz
%SRC \end{enumerate}
%SRC 
%SRC \section{Qux}\label{sec:qux}
%SRC 
%SRC \newtheorem{lemma}{Lemma}
%SRC 
%SRC \begin{lemma}[Qux]\label{thm:qux}
%SRC     Qux
%SRC \end{lemma}
%SRC 
%SRC \end{document}

%TEX main.aux
%SRC \relax
%SRC \@writefile{lof}{\contentsline {figure}{\numberline {1}{\ignorespaces Bar\relax }}{1}\protected@file@percent }
%SRC \providecommand*\caption@xref[2]{\@setref\relax\@undefined{#1}}
%SRC \newlabel{fig:bar}{{1}{1}}
%SRC \@writefile{toc}{\contentsline {section}{\numberline {1}Foo}{1}\protected@file@percent }
%SRC \newlabel{sec:foo}{{1}{1}}
%SRC \newlabel{eq:foo}{{1}{1}}
%SRC \@writefile{toc}{\contentsline {section}{\numberline {2}Bar}{1}\protected@file@percent }
%SRC \newlabel{sec:bar}{{2}{1}}
%SRC \@writefile{toc}{\contentsline {section}{\numberline {3}Baz}{1}\protected@file@percent }
%SRC \newlabel{sec:baz}{{3}{1}}
%SRC \newlabel{itm:foo}{{1}{1}}
%SRC \newlabel{itm:bar}{{2}{1}}
%SRC \newlabel{itm:baz}{{3}{1}}
%SRC \@writefile{toc}{\contentsline {section}{\numberline {4}Qux}{1}\protected@file@percent }
%SRC \newlabel{sec:qux}{{4}{1}}
%SRC \newlabel{thm:qux}{{1}{1}}

%BIB main.bib
%SRC @article{foo,}
%SRC 
%SRC @string{bar = "bar"}"#;

#[test]
fn filter_type_section() -> Result<()> {
    assert_symbols!(find_symbols(FIXTURE, "section")?);
    Ok(())
}

#[test]
fn filter_type_figure() -> Result<()> {
    assert_symbols!(find_symbols(FIXTURE, "figure")?);
    Ok(())
}

#[test]
fn filter_type_item() -> Result<()> {
    assert_symbols!(find_symbols(FIXTURE, "item")?);
    Ok(())
}

#[test]
fn filter_type_math() -> Result<()> {
    assert_symbols!(find_symbols(FIXTURE, "math")?);
    Ok(())
}

#[test]
fn filter_bibtex() -> Result<()> {
    assert_symbols!(find_symbols(FIXTURE, "bibtex")?);
    Ok(())
}
