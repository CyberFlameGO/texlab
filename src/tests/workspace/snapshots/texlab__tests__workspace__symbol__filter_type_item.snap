---
source: src/tests/workspace/symbol.rs
expression: result.actual_symbols
---
[
  {
    "deprecated": false,
    "kind": 22,
    "location": {
      "uri": "[tmp]/main.tex",
      "range": {
        "start": {
          "line": 24,
          "character": 4
        },
        "end": {
          "line": 24,
          "character": 28
        }
      }
    },
    "name": "1"
  },
  {
    "deprecated": false,
    "kind": 22,
    "location": {
      "uri": "[tmp]/main.tex",
      "range": {
        "start": {
          "line": 25,
          "character": 4
        },
        "end": {
          "line": 25,
          "character": 28
        }
      }
    },
    "name": "2"
  },
  {
    "deprecated": false,
    "kind": 22,
    "location": {
      "uri": "[tmp]/main.tex",
      "range": {
        "start": {
          "line": 26,
          "character": 4
        },
        "end": {
          "line": 26,
          "character": 28
        }
      }
    },
    "name": "3"
  }
]
